persist/fs/commitlog/reader.go

```diff
@@ -109,58 +165,171 @@ func (r *reader) Read() (
-		return
-	}
+		if len(entry.Metadata) != 0 {
+			metadataDecoder.Reset(entry.Metadata)
+			decoded, err := metadataDecoder.DecodeLogMetadata()
+			if err != nil {
+				readResponse.resultErr = err
+				outBuf <- readResponse
+				continue
+			}
 
-	series = metadata
-	datapoint = ts.Datapoint{
-		Timestamp: time.Unix(0, entry.Timestamp),
-		Value:     entry.Value,
+			id := r.bytesPool.Get(len(decoded.ID))
+			id.IncRef()
+			id.AppendAll(decoded.ID)
+
+			namespace := r.bytesPool.Get(len(decoded.Namespace))
+			namespace.IncRef()
+			namespace.AppendAll(decoded.Namespace)
+
+			r.metadata.Lock()
+			r.metadata.lookup[entry.Index] = Series{
+				UniqueIndex: entry.Index,
+				ID:          ts.BinaryID(id),
+				Namespace:   ts.BinaryID(namespace),
+				Shard:       decoded.Shard,
+			}
+			waitGroup, ok := r.metadata.wgs[entry.Index]
+			// Another goroutine is blocked waiting for this metadata
+			if ok {
+				waitGroup.Done()
+			}
+			r.metadata.Unlock()
+			namespace.DecRef()
+			id.DecRef()
+		}
+
+		var pendingResult *sync.WaitGroup
+		var metadata Series
+		var hasMetadata bool
+		var ok bool
+
+		// Try cheap read lock first
+		r.metadata.RLock()
+		metadata, hasMetadata = r.metadata.lookup[entry.Index]
+		r.metadata.RUnlock()
+
+		// The required metadata hasn't been processed yet
+		if !hasMetadata {
+			r.metadata.Lock()
+			metadata, hasMetadata = r.metadata.lookup[entry.Index]
+			if !hasMetadata {
+				waitGroup, hasWaitGroup := r.metadata.wgs[entry.Index]
+				if hasWaitGroup {
+					pendingResult = waitGroup
+				} else {
+					pendingResult = &sync.WaitGroup{}
+					pendingResult.Add(1)
+					r.metadata.wgs[entry.Index] = pendingResult
+				}
+			}
+			r.metadata.Unlock()
+		}
+
+		// The required metadata hasn't been processed yet
+		if pendingResult != nil {
+			pendingResult.Wait()
+			r.metadata.RLock()
+			metadata, ok = r.metadata.lookup[entry.Index]
+			r.metadata.RUnlock()
+			// Something went wrong, maybe the reader was closed early
+			if !ok {
+				readResponse.resultErr = errCommitLogReaderMissingLogMetadata
+				outBuf <- readResponse
+				continue
+			}
+		}
+
+		readResponse.series = metadata
+		readResponse.datapoint = ts.Datapoint{
+			Timestamp: time.Unix(0, entry.Timestamp),
+			Value:     entry.Value,
+		}
+		readResponse.unit = xtime.Unit(byte(entry.Unit))
+		// Copy annotation to prevent reference to pooled byte slice
+		if len(entry.Annotation) > 0 {
+			readResponse.annotation = append([]byte(nil), entry.Annotation...)
+		}
+		// DecRef delayed until this point because even after decoding, underlying
+		// byte slice is still referenced by metadata and annotation
+		arg.bytes.DecRef()
+		arg.bytes.Finalize()
+		outBuf <- readResponse
 	}
-	unit = xtime.Unit(byte(entry.Unit))
-	annotation = entry.Annotation
-	return
 }
 
-func (r *reader) readChunk() ([]byte, error) {
+// Caller is responsible for calling DecRef()
+func (r *reader) readChunk() (checked.Bytes, error) {
 	// Read size of message
```

Comment:
just curious - would increasing read perf here (e.g. `mmap`ing the chunk reader) help overall perf at all?

Reply:
I considered that, but I don't think so. At least not yet. As far as I can tell from my profiling all the bottlenecks in the code are around M3TSZ encoding and MSGPACK decoding.

This PR along with the next one makes the commitlog bootstrapper about 5-6x faster just by parallelizing those things.

Might be interesting to look at parallelizing the reader later to if we see that become a bottleneck, but there is some weird complexity there because I think we still want things to come out of the reader in the same order that they're found in the commitlog.

Reply:
fair enough.